File: Cpp/124.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int mm = root->val;
        f(root,mm);
        return mm;
    }

    int f(TreeNode* root, int& mm){
        if(root == NULL){
            return 0;
        }
        int lc = max(f(root->left, mm),0);
        int rc = max(f(root->right, mm),0);

        mm = max(mm, lc+rc+root->val);

        return root->val + max(lc,rc);
    }

};

File: Python/124.py
class Solution:

    def f(self, root, sol):
        if root == None:
            return 0
        
        lsum = self.f(root.left, sol)
        rsum = self.f(root.right, sol)

        sol[0] = max(sol[0], root.val + lsum + rsum,root.val + lsum, root.val + rsum, root.val)
        return max(root.val + max(lsum , rsum), root.val)

    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        sol = [float('-inf')]
        _ = self.f(root, sol)
        return sol[0]
